### ISSUE TITLE
Magics with toplevel await

### ISSUE THREAD

**Issue by mlucool** [ipython/ipython#14157]
Hi,

It would be nice to be able to use top level await with magics, specifically `%time` and `%%time`.

For example:
```
async def foo():
    return 'bar'
# This works
await foo()
# This fails with SyntaxError: 'await' outside function
%time await foo()
```

**Comment by Carreau** [comment]
Aplogies for the late response, I was at a conference last week. It is interesting, I'll see what I can do.

**Comment by Carreau** [comment]
So there are multiple things this will entail, I want to understand some of the consequences, and this could be done in multiple steps. 

1) Allow async magics, that in itself might have the most consequences, I thin if allow that that will make magics in sync context invalid. 

```
def f():
   a = %time 1+1
```   

This will now fail with `SyntaxError: 'await' outside async function`, I need to figure out if we can mitigate this or figure out ahead of time if a magic is/will be sync or async.

Even if we detect the sync async magics, should we make the `time`/`timeit` async or introduce `atime`/`atimeit`, ortherwise we'll fallback back into the above problem.

One possibility is also to try to base magics on ast-rewriting, but that my prevent having multiple magics.

**Comment by bollwyvl** [comment]
> `atime`/`atimeit`

This seems like a fine solution. If this color was knowable from outside (without `inspect`), e.g. `register_line_cell_magic(async=True)`, that would be great, too.

Another typographic approach, once those are known, might be to have The One Async Magic, e.g. `%await timeit` (which could maybe be shorthanded to `%a`)... adding _another_ special symbol would make everybody cry.

**Comment by Carreau** [comment]
> > `atime`/`atimeit`
> 
> This seems like a fine solution. If this color was knowable from outside (without `inspect`), e.g. `register_line_cell_magic(async=True)`, that would be great, too.
> 
> Another typographic approach, once those are known, might be to have The One Async Magic, e.g. `%await timeit` (which could maybe be shorthanded to `%a`)... adding _another_ special symbol would make everybody cry.

Yeah, we can use inspect.iscoroutinefunction() but the problem is not there, the problem is that
1) the input transformer does not know about whcih magics are registered it jsut turn `%foo` into `get_ipython().run_line_magic('foo', ...)` (or await `get_ipython().run_async_line_magic('foo', ...)`, but we don't know if the magic is sync or async.
2) You may want `%time` to magically be sync if it's content is sync (so you can do `def f():\n... x = %time <sync code>`, or async if `async def f(): %time <async code>`). Which you don't know until you actually hit the magic.

But in general this means that every implementation that want to support both sync and async need two definition.

I _think_ I can make it work in most case with something that manipulate the ast. But is leads to issues with all the non-python syntax IPython supports.

**Comment by Carreau** [comment]
So allowing something like ``%%async <magicname>`` assuming `<magicname>` is (or provides) an async function is fairly straitforward.

Having `%%time` (or other magic) works both magically for async and sync, will likely be much more difficult.

It might be possible to make it work in _most_ cases, by having ``%%foo`` be async _by default_, so that things that do not run event loop can work, but and add a ``%%sync foo`` to force sync mode for foo.

I'm thinking:
  - [ ] adding ``%%sync foo`` and ``%%async foo``.  (or ``%%await foo``, not sure which one.)
  - [ ] for now, have ``%%foo`` be an alias of `%%sync foo`
  - [ ] progressively rewrite the magics that could be made async. 
  - [ ] at some point of the future, switch ``%%foo`` to be an alias for ``%%async foo``, and make ``%%sync foo`` to be explicit for code that need to for example excute code like ``asyncio.run``.


It will also be possible for line magics but likely slightly more complicated as we have a few edge case to take care of (assignment, and here I think we can directly handle sync async with `x = await %time ...`, vs `x = %time`.

The big difficulty here is that most magics that want to support async will need to have rewritten, and in some (many?) case each magic that want to support both sync and async need to likely expose two functions, and that might be long term work.

**Comment by mlucool** [comment]
Closing given the current ROI is too low
